From 7285a09bfc3bcca8a72ac795acae302afa8f9f55 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Robert=20L=C3=B6hning?= <robert.loehning@qt.io>
Date: Wed, 18 Jun 2025 16:02:32 +0200
Subject: Replace check for endless recursion when loading

The old check parsed the tree of SvgNodes again and again which lead to
quadratic complexity. Instead, set and check a bool where the recursion
may actually happen which is faster and only has linear complexity.

Partially reverts 0332df304f013ded362537c1f61556098b875352

I chose to have the check in QSvgPattern::renderPattern() because:

- It not only appears in the recursive backtrace of the stack-overflow
  which was fixed using the qudratic check, but also in the backtrace
  of another, still unfixed stack overflow. That way, both can be fixed
  by the same patch. Credit to OSS-Fuzz for finding them.
- The function already had some error checking and returns a default
  value when it cannot render the content. In the same way, I can return
  a QImage of the right size but without any content when the endless
  recursion is about to happen.

[ChangeLog] Speed up loading by replacing check for cyclic elements
[ChangeLog] Fix stack overflow when an element references its child
element using url()

Fixes: QTBUG-137553
Pick-to: 6.10 6.9 6.8
Change-Id: If011c15fde50dcefeb653d1d5995ff1347e7b5ac
Reviewed-by: Hatem ElKharashy <hatem.elkharashy@qt.io>
(cherry picked from commit 9e5bed9584ab65d56cd5fbac0471e06e37a54412)
---
 src/svg/qsvghandler.cpp                      | 3 +--
 src/svg/qsvgstructure.cpp                    | 8 ++++++++
 src/svg/qsvgstructure_p.h                    | 1 +
 tests/auto/qsvgrenderer/tst_qsvgrenderer.cpp | 3 +++
 4 files changed, 13 insertions(+), 2 deletions(-)

diff --git x/qtsvg/src/svg/qsvghandler.cpp y/qtsvg/src/svg/qsvghandler.cpp
index 48f9f0b25c01404398857b6cbf1935a64f0dd861..c871a20cdc722e7feab6dd3273da708140f54f7c 100644
--- x/qtsvg/src/svg/qsvghandler.cpp
+++ y/qtsvg/src/svg/qsvghandler.cpp
@@ -4647,8 +4647,7 @@ void QSvgHandler::parse()
             // this point is to do what everyone else seems to do and
             // ignore the reported namespaceUri completely.
             if (remainingUnfinishedElements
-                    && startElement(xml->name().toString(), xml->attributes())
-                    && !detectCyclesAndWarn(m_doc)) {
+                    && startElement(xml->name().toString(), xml->attributes())) {
                 --remainingUnfinishedElements;
             } else {
                 delete m_doc;
diff --git x/qtsvg/src/svg/qsvgstructure.cpp y/qtsvg/src/svg/qsvgstructure.cpp
index e08f1bc16f0d28281a2b0c382711edf8e030edb7..d073b1c4a04a377dff99ebbd5f22d3d4fc299ea4 100644
--- x/qtsvg/src/svg/qsvgstructure.cpp
+++ y/qtsvg/src/svg/qsvgstructure.cpp
@@ -800,6 +800,7 @@ QSvgPattern::QSvgPattern(QSvgNode *parent, QSvgRectF bounds, QRectF viewBox,
     m_rect(bounds),
     m_viewBox(viewBox),
     m_contentUnits(contentUnits),
+    m_isRendering(false),
     m_transform(transform)
 
 {
@@ -885,6 +886,13 @@ QImage QSvgPattern::renderPattern(QSize size, qreal contentScaleX, qreal content
     }
     pattern.fill(Qt::transparent);
 
+    if (m_isRendering) {
+        qCWarning(lcSvgDraw) << "The pattern is trying to render itself recursively. "
+                                "Returning a transparent QImage of the right size.";
+        return pattern;
+    }
+    QScopedValueRollback<bool> guard(m_isRendering, true);
+
     // Draw the pattern using our QPainter.
     QPainter patternPainter(&pattern);
     QSvgExtraStates patternStates;
diff --git x/qtsvg/src/svg/qsvgstructure_p.h y/qtsvg/src/svg/qsvgstructure_p.h
index 63a839641db6b51442781688ad2c2713f63847f7..f95508f49c4561e847ce50b9c7a8d088935d5cb6 100644
--- x/qtsvg/src/svg/qsvgstructure_p.h
+++ y/qtsvg/src/svg/qsvgstructure_p.h
@@ -241,6 +241,7 @@ private:
     QSvgRectF m_rect;
     QRectF m_viewBox;
     QtSvg::UnitTypes m_contentUnits;
+    mutable bool m_isRendering;
     QTransform m_transform;
 };
 
diff --git x/qtsvg/tests/auto/qsvgrenderer/tst_qsvgrenderer.cpp y/qtsvg/tests/auto/qsvgrenderer/tst_qsvgrenderer.cpp
index b212ad39a77e5ebd5b6690723c664a208f7068ce..56f6664328ead7c390d2b0c724566c8160bdcd8e 100644
--- x/qtsvg/tests/auto/qsvgrenderer/tst_qsvgrenderer.cpp
+++ y/qtsvg/tests/auto/qsvgrenderer/tst_qsvgrenderer.cpp
@@ -1809,6 +1809,9 @@ void tst_QSvgRenderer::ossFuzzLoad_data()
     // resulted in stack overflow
     QTest::newRow("cyclic-reference") // id=42532991
             << R"-(<svg><pattern height="3" width="9" id="c"><path d="v4T1-" stroke="url(#c)"><symbol>)-"_ba;
+    // resulted in stack overflow
+    QTest::newRow("cyclic-reference-from-parent") // id=390467765
+            << R"-(<svg stroke="url(#c)"><pattern height="2" width="4" id="c"/><path stroke="#F00" d="v2"/></svg>)-"_ba;
 }
 
 void tst_QSvgRenderer::ossFuzzLoad()
